// @tarojs/plugin-platform-harmony-hybrid/dist/api/apis/cloud/index.js
/**
 * 云开发 SDK 实例
 *
 * @canUse cloud
 * @null_implementation
 */
class cloud {
    constructor() {
        this.init = (config) => {
            if (config && typeof config === 'object') {
                // do nothing
            }
            else {
                // do nothing
            }
            Promise.resolve();
        };
        this.CloudID = (cloudID) => {
            if (typeof cloudID === 'string') {
                // do nothing
            }
        };
        // @ts-ignore
        this.callFunction = (param) => {
            if (typeof param === 'object') {
                // do nothing
            }
        };
        // @ts-ignore
        this.uploadFile = (param) => {
            if (typeof param === 'object') {
                // do nothing
            }
        };
        // @ts-ignore
        this.downloadFile = (param) => {
            if (typeof param === 'object') {
                // do nothing
            }
        };
        // @ts-ignore
        this.getTempFileURL = (param) => {
            if (typeof param === 'object') {
                // do nothing
            }
        };
        // @ts-ignore
        this.deleteFile = (param) => {
            if (typeof param === 'object') {
                // do nothing
            }
        };
        // @ts-ignore
        this.database = (config) => {
            if (config && typeof config === 'object') {
                // do nothing
            }
        };
        // @ts-ignore
        this.callContainer = (params) => {
            if (typeof params === 'object') {
                // do nothing
            }
        };
    }
}
/**
 * 云开发 SDK 数据库实例 Taro.cloud.database()
 *
 * @canNotUse DB
 */

export { cloud };
