// @tarojs/plugin-platform-harmony-hybrid/dist/api/apis/media/image/chooseImage.js
import { __awaiter } from '../../../../node_modules/.pnpm/@rollup_plugin-typescript@11.1.6_rollup@3.29.5_tslib@2.8.1_typescript@5.4.5/node_modules/tslib/tslib.es6.js';
import { showActionSheet } from '@tarojs/taro-h5';
import native from '../../NativeApi.js';
import { shouldBeObject } from '../../utils/index.js';
import { MethodHandler } from '../../utils/handler.js';

/**
 * 从本地相册选择图片或使用相机拍照。
 *
 * @canUse chooseImage
 * @__object [count, sizeType[original, compressed], sourceType[album, camera]]
 * @__success [tempFilePaths, tempFiles]
 */
const chooseImage = (options) => __awaiter(void 0, void 0, void 0, function* () {
    const name = 'chooseImage';
    // options must be an Object
    const isValid = shouldBeObject(options).flag || typeof options === 'undefined';
    if (!isValid) {
        const res = { errMsg: `${name}:fail invalid params` };
        console.error(res.errMsg);
        return Promise.reject(res);
    }
    const { count = 9, sourceType = ['album', 'camera'], sizeType = ['original', 'compressed'], success, fail, } = options;
    const mediaType = ['image'];
    const handle = new MethodHandler({ name, success, fail });
    let sourceSelected;
    if (sourceType.length === 1) {
        sourceSelected = sourceType[0];
    }
    else if (typeof sourceType !== 'object' || (sourceType.includes('album') && sourceType.includes('camera'))) {
        const selected = yield showActionSheet({ itemList: ['拍摄', '从相册选择'] }).then((res) => {
            sourceSelected = res.tapIndex === 0 ? 'camera' : 'album';
            return true;
        }, () => {
            return false;
        });
        if (!selected) {
            return handle.fail({ errMsg: 'fail cancel' });
        }
    }
    return new Promise((resolve, reject) => {
        native.chooseMediumAssets({
            count: count,
            mediaType: mediaType,
            sourceType: sourceSelected,
            sizeType: sizeType,
            apiName: name,
            success: (res) => {
                const tempFiles = [];
                for (const file of res.tempFiles) {
                    const fileInfo = {
                        path: file.tempFilePath,
                        size: file.size,
                        type: file.tempFilePath.split('.').pop(),
                    };
                    tempFiles.push(fileInfo);
                }
                const result = {
                    tempFilePaths: res.tempFilePaths,
                    tempFiles: tempFiles,
                    errMsg: res.errMsg,
                };
                handle.success(result, { resolve, reject });
            },
            fail: (err) => {
                handle.fail(err, { resolve, reject });
            },
        });
    });
});

export { chooseImage };
