// @tarojs/plugin-platform-harmony-hybrid/dist/api/apis/media/image/getImageInfo.js
import native from '../../NativeApi.js';
import { shouldBeObject } from '../../utils/index.js';
import { MethodHandler } from '../../utils/handler.js';

/**
 * 从本地相册选择图片或使用相机拍照。
 *
 * @canUse getImageInfo
 * @__object [src]
 * @__success
 * [height, orientation[up, up-mirrored, down, down-mirrored, left-mirrored, right, right-mirrored, left],\
 * path, type, width]
 */
const getImageInfo = function (options) {
    const name = 'getImageInfo';
    // options must be an Object
    const isObject = shouldBeObject(options);
    if (!isObject.flag) {
        const res = { errMsg: `getImageInfo:fail ${isObject.msg}` };
        console.error(res.errMsg);
        return Promise.reject(res);
    }
    const { src, success, fail, complete } = options;
    const handle = new MethodHandler({ name, success, fail, complete });
    return new Promise((resolve, reject) => {
        native.getImageInfo({
            src: src,
            success: (res) => {
                const result = {
                    height: res.height,
                    orientation: res.orientation,
                    path: res.path,
                    type: res.type,
                    width: res.width,
                    errMsg: res.errMsg,
                };
                handle.success(result, { resolve, reject });
            },
            fail: (err) => {
                handle.fail(err, { resolve, reject });
            },
        });
    });
};

export { getImageInfo };
