// @tailwindcss-mangle/config/dist/index.js
// src/config.ts
import fs from "fs-extra";
import path from "pathe";

// src/constants.ts
var CONFIG_NAME = "tailwindcss-mangle";

// src/defaults.ts
import process from "process";
import { defaultMangleClassFilter } from "@tailwindcss-mangle/shared";
import { CSS_LANGS_RE } from "is-css-request";
var defaultPipelineInclude = [
  /\.(html|js|cjs|mjs|ts|cts|mts|jsx|tsx|vue|svelte|astro|elm|php|phtml|mdx|md)(?:$|\?)/,
  CSS_LANGS_RE
];
var defaultPipelineExclude = [];
function getDefaultRegistryConfig() {
  return {
    output: {
      file: ".tw-patch/tw-class-list.json",
      stripUniversalSelector: true,
      pretty: true
    },
    tailwind: {}
  };
}
function getDefaultTransformerConfig() {
  return {
    filter: defaultMangleClassFilter,
    sources: {
      include: defaultPipelineInclude,
      exclude: defaultPipelineExclude
    },
    disabled: process.env.NODE_ENV === "development",
    registry: {
      file: ".tw-patch/tw-class-list.json",
      mapping: {
        enabled: false,
        file: ".tw-patch/tw-map-list.json",
        loose: true
      }
    },
    preserve: {
      functions: [],
      classes: []
    }
  };
}
function getDefaultUserConfig() {
  return {
    registry: getDefaultRegistryConfig(),
    transformer: getDefaultTransformerConfig()
  };
}

// src/config.ts
var c12Promise;
async function loadC12() {
  if (!c12Promise) {
    c12Promise = import("c12");
  }
  return c12Promise;
}
async function getConfig(cwd) {
  const { loadConfig } = await loadC12();
  return loadConfig({
    name: CONFIG_NAME,
    defaults: {
      ...getDefaultUserConfig()
    },
    cwd
  });
}
var defineConfig = (config) => config;
function initConfig(cwd) {
  return fs.outputFile(
    path.resolve(cwd, `${CONFIG_NAME}.config.ts`),
    `import { defineConfig } from 'tailwindcss-patch'

export default defineConfig({
  registry: {
    output: {
      file: '.tw-patch/tw-class-list.json',
    },
  },
  transformer: {
    registry: {
      file: '.tw-patch/tw-class-list.json',
    },
  },
})
`,
    "utf8"
  );
}
export {
  CONFIG_NAME,
  defineConfig,
  getConfig,
  getDefaultRegistryConfig,
  getDefaultTransformerConfig,
  getDefaultUserConfig,
  initConfig
};
