// @tarojs/plugin-platform-harmony-hybrid/dist/api/apis/location/chooseLocation.js
import { shouldBeObject } from '../utils/index.js';
import { MethodHandler } from '../utils/handler.js';
import nativeLocation from './NativeLocation.js';

/**
 * 打开地图选择位置。
 *
 * @canUse chooseLocation
 * @__object [latitude, longitude]
 * @__success [address, latitude, longitude, name]
 */
const chooseLocation = (options = {}) => {
    const name = 'chooseLocation';
    const isObject = shouldBeObject(options);
    if (!isObject.flag) {
        const res = { errMsg: `openLocation:fail ${isObject.msg}` };
        return Promise.reject(res);
    }
    const { latitude, longitude, success, fail, complete } = options;
    const handle = new MethodHandler({ name, success, fail, complete });
    return new Promise((resolve, reject) => {
        nativeLocation.chooseLocation({
            latitude,
            longitude,
            success: (res) => {
                const result = {
                    address: res.address,
                    latitude: res.location.latitude,
                    longitude: res.location.longitude,
                    name: res.name,
                    errMsg: `${name}:ok`
                };
                handle.success(result, { resolve, reject });
            },
            fail: (res) => {
                handle.fail(res, { resolve, reject });
            }
        });
    });
};

export { chooseLocation };
