// @tarojs/plugin-platform-harmony-hybrid/dist/api/apis/device/screen/setKeepScreenOn.js
import native from '../../NativeApi.js';
import { shouldBeObject, getParameterError } from '../../utils/index.js';
import { MethodHandler } from '../../utils/handler.js';

/**
 * 设置是否保持常亮状态
 *
 * @canUse setKeepScreenOn
 * @__object [keepScreenOn]
 */
const setKeepScreenOn = (options) => {
    const name = 'setKeepScreenOn';
    return new Promise((resolve, reject) => {
        // options must be an Object
        const isObject = shouldBeObject(options);
        if (!isObject.flag) {
            const res = { errMsg: `${name}:fail ${isObject.msg}` };
            console.error(res.errMsg);
            return reject(res);
        }
        const { keepScreenOn, success, fail, complete } = options;
        const handle = new MethodHandler({ name, success, fail, complete });
        if (typeof keepScreenOn !== 'boolean') {
            return handle.fail({
                errMsg: getParameterError({
                    para: 'value',
                    correct: 'boolean',
                    wrong: keepScreenOn,
                }),
            }, { resolve, reject });
        }
        native.setKeepScreenOn({
            keepScreenOn: keepScreenOn,
            success: (res) => {
                handle.success(res, { resolve, reject });
            },
            fail: (err) => {
                handle.fail(err, { resolve, reject });
            },
        });
    });
};

export { setKeepScreenOn };
