// @tarojs/plugin-platform-harmony-hybrid/dist/api/apis/device/screen/setScreenBrightness.js
import { shouldBeObject, getParameterError } from '../../utils/index.js';
import { MethodHandler } from '../../utils/handler.js';
import native from '../../NativeApi.js';

/**
 * 设置屏幕亮度
 *
 * @canUse setScreenBrightness
 * @__object [value]
 */
const setScreenBrightness = (options) => {
    const name = 'setScreenBrightness';
    return new Promise((resolve, reject) => {
        // options must be an Object
        const isObject = shouldBeObject(options);
        if (!isObject.flag) {
            const res = { errMsg: `${name}:fail ${isObject.msg}` };
            console.error(res.errMsg);
            return reject(res);
        }
        const { value, success, fail, complete } = options;
        const handle = new MethodHandler({ name, success, fail, complete });
        if (typeof value !== 'number') {
            return handle.fail({
                errMsg: getParameterError({
                    para: 'value',
                    correct: 'number',
                    wrong: value,
                }),
            }, { resolve, reject });
        }
        // @ts-ignore
        native.setScreenBrightness({
            value: value,
            success: (res) => {
                handle.success(res, { resolve, reject });
            },
            fail: (err) => {
                handle.fail(err, { resolve, reject });
            },
        });
    });
};

export { setScreenBrightness };
