// @tarojs/plugin-platform-harmony-hybrid/dist/components/components-react/Map/Map.js
import { jsx } from 'react/jsx-runtime';
import React from 'react';
import sameLayerRender, { clearJsObj } from '../SameLayerRender.js';

class HosMap extends React.Component {
    constructor(props) {
        super(props);
        this.componentId = `HosMap_${Math.floor(Math.random() * 100000)}_${Date.now()}`;
    }
    componentDidMount() {
        const { latitude, longitude, scale, rotate, skew, minScale, maxScale, enableZoom, enableScroll, enableRotate, showCompass, showScale, showLocation, enableTraffic, markers, onTap, onMarkerTap, onCalloutTap, onUpdated, onRegionChange, onPoiTap, onPolylineTap, onError, onAnchorPointTap, } = this.props;
        this.nativeRenderArgs = {
            componentId: this.componentId,
            latitude,
            longitude,
            zoom: scale,
            bearing: rotate,
            tilt: skew,
            minZoom: minScale,
            maxZoom: maxScale,
            zoomGesturesEnabled: enableZoom,
            scrollGesturesEnabled: enableScroll,
            rotateGesturesEnabled: enableRotate,
            compassControlsEnabled: showCompass,
            scaleControlsEnabled: showScale,
            setMyLocationEnabled: showLocation,
            setTrafficEnabled: enableTraffic,
            markers,
            onMapClick: onTap,
            onMarkerClick: onMarkerTap,
            onBubbleClick: onCalloutTap,
            onMapLoad: onUpdated,
            onCameraChange: onRegionChange,
            onPoiClick: onPoiTap,
            onPolylineClick: onPolylineTap,
            onError: onError,
            onMyLocationButtonClick: onAnchorPointTap,
        };
        sameLayerRender.transferSameLayerArgs(this.nativeRenderArgs);
    }
    componentWillUnmount() {
        // 释放JS侧存储的渲染参数
        clearJsObj(this.nativeRenderArgs);
    }
    render() {
        const { style, className, id } = this.props;
        return (jsx("div", { className: className, style: style, id: id, children: jsx("embed", { style: { width: '100%', height: '100%', objectPosition: 'inherit', display: 'block' }, id: this.componentId, type: "native/hos-map" }) }));
    }
}

export { HosMap as default };
