// @tarojs/plugin-platform-harmony-hybrid/dist/api/apis/device/bluetooth/getBluetoothAdapterState.js
import { shouldBeObject } from '../../utils/index.js';
import { MethodHandler } from '../../utils/handler.js';
import native from '../../NativeApi.js';

/**
 * 获取本机蓝牙适配器状态
 *
 * @canUse getBluetoothAdapterState
 * @__success [available, discovering]
 */
const getBluetoothAdapterState = (options) => {
    const name = 'getBluetoothAdapterState';
    const isValid = shouldBeObject(options).flag || typeof options === 'undefined';
    if (!isValid) {
        const res = { errMsg: `${name}:fail invalid params` };
        console.error(res.errMsg);
        return Promise.reject(res);
    }
    const { success, fail, complete } = options || {};
    const handle = new MethodHandler({ name, success, fail, complete });
    return new Promise((resolve, reject) => {
        native.getBluetoothAdapterState({
            success: (res) => {
                const result = {
                    available: res[0].available,
                    discovering: res[0].discovering,
                    errMsg: '',
                };
                handle.success(result, { resolve, reject });
            },
            fail: (err) => {
                handle.fail(err, { resolve, reject });
            },
        });
    });
};

export { getBluetoothAdapterState };
