// @tarojs/plugin-platform-harmony-hybrid/dist/api/apis/media/index.js
export { chooseMessageFile, createLivePusherContext, createMediaAudioPlayer, createMediaContainer, createMediaRecorder, createVideoContext, createVideoDecoder, createWebAudioContext, cropImage, editImage, exitVoIPChat, getAvailableAudioSources, getBackgroundAudioManager, getBackgroundAudioPlayerState, join1v1Chat, joinVoIPChat, offVoIPChatInterrupted, offVoIPChatMembersChanged, offVoIPChatSpeakersChanged, offVoIPChatStateChanged, offVoIPVideoMembersChanged, onBackgroundAudioPause, onBackgroundAudioPlay, onBackgroundAudioStop, onVoIPChatInterrupted, onVoIPChatMembersChanged, onVoIPChatSpeakersChanged, onVoIPChatStateChanged, onVoIPVideoMembersChanged, pauseBackgroundAudio, pauseVoice, playBackgroundAudio, playVoice, seekBackgroundAudio, setEnable1v1Chat, setInnerAudioOption, startRecord, stopBackgroundAudio, stopRecord, stopVoice, subscribeVoIPVideoMembers, updateVoIPChatMuteConfig } from '@tarojs/taro-h5';
export { createAudioContext, createInnerAudioContext } from './audio/index.js';
import './background-audio/index.js';
export { createCameraContext } from './camera.js';
import './image/index.js';
export { createLivePlayerContext } from './live/index.js';
export { createMapContext } from './map/index.js';
import './media-recorder.js';
export { getRecorderManager } from './recorder.js';
import './video/index.js';
import './video-decoder.js';
import './video-processing.js';
import './voip.js';

/**
 * 创建 live-pusher 上下文 LivePusherContext 对象
 *
 * @canNotUse createLivePusherContext
 */
