// @tarojs/plugin-generator/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const inquirer = __importStar(require("inquirer"));
const es5_1 = require("./generators/es5");
const tailwindcss_1 = require("./generators/tailwindcss");
const error_1 = require("./utils/error");
/**
 * 命令行扩展
 */
exports.default = (ctx) => {
    ctx.registerCommand({
        name: 'new',
        // 命令钩子
        async fn() {
            const choices = {
                tailwindcss: 'tailwindcss',
                es5: 'es5',
            };
            const answer = await inquirer.prompt({
                type: 'list',
                name: 'choice',
                message: '启用可选功能',
                choices: [
                    { name: '启用「Tailwind CSS」支持', value: choices.tailwindcss },
                    { name: '启用「编译为 ES5」', value: choices.es5 },
                ],
            });
            switch (answer.choice) {
                case choices.tailwindcss: {
                    await (0, error_1.safely)(() => (0, tailwindcss_1.tailwindcssGenerator)(ctx));
                    break;
                }
                case choices.es5: {
                    await (0, error_1.safely)(() => (0, es5_1.es5Generator)(ctx));
                    break;
                }
                default: {
                    break;
                }
            }
        },
    });
};
