// @tarojs/plugin-platform-harmony-hybrid/dist/api/apis/media/audio/AudioContext.js
import Taro from '@tarojs/api';

/**
 * AudioContext 实例
 *
 * @canUse AudioContext
 * @__class [pause, play, seek, setSrc]
 */
class AudioContext {
    constructor() {
        this.play = () => { var _a; return (_a = this.Instance) === null || _a === void 0 ? void 0 : _a.play(); };
        this.pause = () => { var _a; return (_a = this.Instance) === null || _a === void 0 ? void 0 : _a.pause(); };
        this.stop = () => {
            this.pause();
            this.seek(0);
        };
        this.seek = (position) => {
            if (this.Instance) {
                this.Instance.currentTime = position;
            }
        };
        this.setSrc = (src) => {
            if (this.Instance) {
                this.Instance.src = src;
            }
        };
        this.Instance = new Audio();
        Taro.eventCenter.on('__taroRouterChange', () => {
            this.stop();
        });
    }
    get paused() {
        var _a, _b;
        return (_b = (_a = this.Instance) === null || _a === void 0 ? void 0 : _a.paused) !== null && _b !== void 0 ? _b : true;
    }
    get src() {
        var _a;
        return ((_a = this.Instance) === null || _a === void 0 ? void 0 : _a.src) || '';
    }
    set src(e) {
        this.setProperty('src', e);
    }
    setProperty(key, value) {
        if (this.Instance) {
            this.Instance[key] = value;
        }
    }
}

export { AudioContext };
