// @tarojs/plugin-platform-harmony-hybrid/dist/api/index.js
import taro from './apis/taro.js';
export { Behavior, Current, ENV_TYPE, Events, Link, canIUseWebp, eventCenter, getEnv, history, initPxTransform, interceptorify, interceptors, options, preload, pxTransform, requirePlugin } from './apis/taro.js';
import './apis/index.js';
export { addCard, addFileToFavorites, addPhoneCalendar, addPhoneRepeatCalendar, addRecentColorSign, addVideoToFavorites, advancedGeneralIdentify, animalClassify, applyAddToMyApps, arrayBufferToBase64, authPrivateMessage, authorize, base64ToArrayBuffer, batchGetStorage, batchGetStorageSync, batchSetStorage, batchSetStorageSync, checkIsAddedToMyMiniProgram, checkIsOpenAccessibility, checkIsPictureInPictureActive, checkIsSupportFacialRecognition, checkSession, chooseAddress, chooseContact, chooseInvoice, chooseInvoiceTitle, chooseLicensePlate, chooseMessageFile, choosePoi, closeSocket, connectSocket, connectWifi, createAnimation, createBLEPeripheralServer, createInferenceSession, createLivePusherContext, createMediaAudioPlayer, createMediaContainer, createMediaQueryObserver, createMediaRecorder, createOffscreenCanvas, createRewardedVideoAd, createTCPSocket, createUDPSocket, createVKSession, createVideoContext, createVideoDecoder, createWebAudioContext, createWorker, cropImage, dishClassify, editImage, exitVoIPChat, faceDetect, faceVerifyForPay, getAvailableAudioSources, getBLEMTU, getBackgroundAudioManager, getBackgroundAudioPlayerState, getBackgroundFetchToken, getBatteryInfo, getBatteryInfoSync, getBeacons, getChannelsLiveInfo, getChannelsLiveNoticeInfo, getChannelsShareKey, getClipboardData, getDeviceVoIPList, getEnterOptionsSync, getExtConfig, getExtConfigSync, getGroupEnterInfo, getGuildInfo, getHCEState, getInferenceEnvInfo, getLocalIPAddress, getNFCAdapter, getNetworkType, getOpenUserInfo, getPerformance, getPrivacySetting, getQQRunData, getRandomValues, getRendererUserAgent, getScreenRecordingState, getSelectedTextRange, getSkylineInfo, getSkylineInfoSync, getSwanId, getUserCryptoManager, getUserProfile, getWeRunData, getWifiList, hideLoading, hideTabBar, hideTabBarRedDot, hideToast, imageAudit, initFaceDetect, initTabBarApis, isAddedToMyApps, isBluetoothDevicePaired, isVKSupport, join1v1Chat, joinVoIPChat, loadFontFace, logoClassify, makeBluetoothPair, navigateBack, navigateBackMiniProgram, navigateBackSmartProgram, navigateToMiniProgram, navigateToSmartGameProgram, navigateToSmartProgram, objectDetectIdentify, ocrBankCard, ocrDrivingLicense, ocrIdCard, ocrVehicleLicense, offAppHide, offAppShow, offAudioInterruptionBegin, offAudioInterruptionEnd, offBLECharacteristicValueChange, offBLEConnectionStateChange, offBLEMTUChange, offBLEPeripheralConnectionStateChanged, offBeaconServiceChange, offBeaconUpdate, offCopyUrl, offDeviceMotionChange, offError, offGyroscopeChange, offHCEMessage, offLocalServiceDiscoveryStop, offLocalServiceFound, offLocalServiceLost, offLocalServiceResolveFail, offNetworkStatusChange, offNetworkWeakChange, offPageNotFound, offScreenRecordingStateChanged, offThemeChange, offUnhandledRejection, offVoIPChatInterrupted, offVoIPChatMembersChanged, offVoIPChatSpeakersChanged, offVoIPChatStateChanged, offVoIPVideoMembersChanged, offWindowResize, onAppHide, onAppShow, onAudioInterruptionBegin, onAudioInterruptionEnd, onBLEMTUChange, onBLEPeripheralConnectionStateChanged, onBackgroundAudioPause, onBackgroundAudioPlay, onBackgroundAudioStop, onBackgroundFetchData, onBeaconServiceChange, onBeaconUpdate, onCopyUrl, onDeviceMotionChange, onError, onHCEMessage, onLocalServiceDiscoveryStop, onLocalServiceFound, onLocalServiceLost, onLocalServiceResolveFail, onNetworkStatusChange, onNetworkWeakChange, onPageNotFound, onScreenRecordingStateChanged, onSocketClose, onSocketError, onSocketMessage, onSocketOpen, onThemeChange, onUnhandledRejection, onVoIPChatInterrupted, onVoIPChatMembersChanged, onVoIPChatSpeakersChanged, onVoIPChatStateChanged, onVoIPVideoMembersChanged, onWindowResize, openBusinessView, openCard, openChannelsActivity, openChannelsEvent, openChannelsLive, openCustomerServiceChat, openEmbeddedMiniProgram, openPrivacyContract, openQzonePublish, openSystemBluetoothSetting, pageScrollTo, pauseBackgroundAudio, pauseVoice, plantClassify, playBackgroundAudio, playVoice, pluginLogin, preloadAssets, preloadSkylineView, preloadSubPackage, preloadWebview, reLaunch, redirectTo, removeTabBarBadge, reportEvent, reportMonitor, reportPerformance, requestDeviceVoIP, requestOrderPayment, requestPolymerPayment, requestSubscribeDeviceMessage, requirePrivacyAuthorize, reserveChannelsLive, saveFileToDisk, seekBackgroundAudio, sendHCEMessage, sendSms, sendSocketMessage, setBackgroundColor, setBackgroundTextStyle, setClipboardData, setCustomDress, setEnable1v1Chat, setInnerAudioOption, setOfficialDress, setPageInfo, setTabBarBadge, setTabBarItem, setTabBarStyle, setVisualEffectOnCapture, setWifiList, setWindowSize, shareFileMessage, shareToWeRun, shareVideoMessage, showActionSheet, showLoading, showModal, showRedPackage, showShareImageMenu, showTabBar, showTabBarRedDot, showToast, startBeaconDiscovery, startDeviceMotionListening, startFacialRecognitionVerify, startFacialRecognitionVerifyAndUploadVideo, startHCE, startLocalServiceDiscovery, startPullDownRefresh, startRecord, startWifi, stopBackgroundAudio, stopBeaconDiscovery, stopDeviceMotionListening, stopFaceDetect, stopHCE, stopLocalServiceDiscovery, stopPullDownRefresh, stopRecord, stopVoice, stopWifi, subscribeVoIPVideoMembers, switchTab, textReview, textToAudio, tradePay, updateQQApp, updateShareMenu, updateVoIPChatMuteConfig, updateWeChatApp } from '@tarojs/taro-h5';
export { createInterstitialAd } from './apis/ad/index.js';
export { setEnableDebug } from './apis/base/debug/setEnableDebug.js';
export { getRealtimeLogManager } from './apis/base/debug/getRealtimeLogManager.js';
export { getLogManager } from './apis/base/debug/getLogManager.js';
export { getAppAuthorizeSetting } from './apis/base/getAppAuthorizeSetting/index.js';
export { getSystemSetting } from './apis/base/getSystemSetting/index.js';
export { getAppBaseInfo, getDeviceInfo, getSystemInfo, getSystemInfoAsync, getSystemInfoSync, getWindowInfo, openAppAuthorizeSetting } from './apis/base/system.js';
export { getUpdateManager } from './apis/base/update/getUpdateManager.js';
export { getLaunchOptionsSync, initLaunchOptions } from './apis/base/weapp/life-cycle.js';
export { env } from './apis/base/index.js';
export { createCanvasContext } from './apis/canvas/createCanvasContext.js';
export { canvasToTempFilePath } from './apis/canvas/canvasToTempFilePath.js';
export { canvasPutImageData } from './apis/canvas/canvasPutImageData.js';
export { canvasGetImageData } from './apis/canvas/canvasGetImageData.js';
export { cloud } from './apis/cloud/index.js';
export { getExptInfoSync } from './apis/data-analysis/getExptInfoSync.js';
export { reportAnalytics } from './apis/data-analysis/reportAnalytics.js';
export { offAccelerometerChange } from './apis/device/accelerometer/offAccelerometerChange.js';
export { onAccelerometerChange } from './apis/device/accelerometer/onAccelerometerChange.js';
export { startAccelerometer } from './apis/device/accelerometer/startAccelerometer.js';
export { stopAccelerometer } from './apis/device/accelerometer/stopAccelerometer.js';
export { closeBluetoothAdapter } from './apis/device/bluetooth/closeBluetoothAdapter.js';
export { getBluetoothAdapterState } from './apis/device/bluetooth/getBluetoothAdapterState.js';
export { getBluetoothDevices } from './apis/device/bluetooth/getBluetoothDevices.js';
export { getConnectedBluetoothDevices } from './apis/device/bluetooth/getConnectedBluetoothDevices.js';
export { offBluetoothAdapterStateChange } from './apis/device/bluetooth/offBluetoothAdapterStateChange.js';
export { offBluetoothDeviceFound } from './apis/device/bluetooth/offBluetoothDeviceFound.js';
export { onBluetoothAdapterStateChange } from './apis/device/bluetooth/onBluetoothAdapterStateChange.js';
export { onBluetoothDeviceFound } from './apis/device/bluetooth/onBluetoothDeviceFound.js';
export { openBluetoothAdapter } from './apis/device/bluetooth/openBluetoothAdapter.js';
export { startBluetoothDevicesDiscovery } from './apis/device/bluetooth/startBluetoothDevicesDiscovery.js';
export { stopBluetoothDevicesDiscovery } from './apis/device/bluetooth/stopBluetoothDevicesDiscovery.js';
export { closeBLEConnection } from './apis/device/bluetooth-ble/closeBLEConnection.js';
export { createBLEConnection } from './apis/device/bluetooth-ble/createBLEConnection.js';
export { getBLEDeviceCharacteristics } from './apis/device/bluetooth-ble/getBLEDeviceCharacteristics.js';
export { getBLEDeviceRSSI } from './apis/device/bluetooth-ble/getBLEDeviceRSSI.js';
export { getBLEDeviceServices } from './apis/device/bluetooth-ble/getBLEDeviceServices.js';
export { notifyBLECharacteristicValueChange } from './apis/device/bluetooth-ble/notifyBLECharacteristicValueChange.js';
export { onBLECharacteristicValueChange } from './apis/device/bluetooth-ble/onBLECharacteristicValueChange.js';
export { onBLEConnectionStateChange } from './apis/device/bluetooth-ble/onBLEConnectionStateChange.js';
export { readBLECharacteristicValue } from './apis/device/bluetooth-ble/readBLECharacteristicValue.js';
export { setBLEMTU } from './apis/device/bluetooth-ble/setBLEMTU.js';
export { writeBLECharacteristicValue } from './apis/device/bluetooth-ble/writeBLECharacteristicValue.js';
export { offCompassChange } from './apis/device/compass/offCompassChange.js';
export { onCompassChange } from './apis/device/compass/onCompassChange.js';
export { startCompass } from './apis/device/compass/startCompass.js';
export { stopCompass } from './apis/device/compass/stopCompass.js';
export { addPhoneContact } from './apis/device/contact/addPhoneContact.js';
export { onGyroscopeChange } from './apis/device/gyroscope/onGyroscopeChange.js';
export { startGyroscope } from './apis/device/gyroscope/startGyroscope.js';
export { stopGyroscope } from './apis/device/gyroscope/stopGyroscope.js';
export { hideKeyboard } from './apis/device/keyBoard/hideKeyboard/index.js';
export { offKeyboardHeightChange } from './apis/device/keyBoard/offKeyboardHeightChange.js';
export { onKeyboardHeightChange } from './apis/device/keyBoard/onKeyboardHeightChange.js';
export { offMemoryWarning } from './apis/device/memory/offMemoryWarning.js';
export { onMemoryWarning } from './apis/device/memory/onMemoryWarning.js';
export { makePhoneCall } from './apis/device/phone/index.js';
export { scanCode } from './apis/device/scan.js';
export { getScreenBrightness } from './apis/device/screen/getScreenBrightness.js';
export { offUserCaptureScreen } from './apis/device/screen/offUserCaptureScreen.js';
export { onUserCaptureScreen } from './apis/device/screen/onUserCaptureScreen.js';
export { setKeepScreenOn } from './apis/device/screen/setKeepScreenOn.js';
export { setScreenBrightness } from './apis/device/screen/setScreenBrightness.js';
export { vibrateLong, vibrateShort } from './apis/device/vibrate.js';
export { getConnectedWifi } from './apis/device/wifi/getConnectedWifi.js';
export { offGetWifiList } from './apis/device/wifi/offGetWifiList.js';
export { offWifiConnected } from './apis/device/wifi/offWifiConnected.js';
export { offWifiConnectedWithPartialInfo } from './apis/device/wifi/offWifiConnectedWithPartialInfo.js';
export { onGetWifiList } from './apis/device/wifi/onGetWifiList.js';
export { onWifiConnected } from './apis/device/wifi/onWifiConnected.js';
export { onWifiConnectedWithPartialInfo } from './apis/device/wifi/onWifiConnectedWithPartialInfo.js';
export { getFileInfo } from './apis/files/getFileInfo.js';
export { getFileSystemManager } from './apis/files/getFileSystemManager.js';
export { getSavedFileInfo } from './apis/files/getSavedFileInfo.js';
export { getSavedFileList } from './apis/files/getSavedFileList.js';
export { openDocument } from './apis/files/openDocument.js';
export { removeSavedFile } from './apis/files/removeSavedFile.js';
export { saveFile } from './apis/files/saveFile.js';
export { getCurrentPages } from '@tarojs/router';
export { getApp, getCurrentInstance } from './apis/framework/index.js';
export { chooseLocation } from './apis/location/chooseLocation.js';
export { getFuzzyLocation } from './apis/location/getFuzzyLocation.js';
export { getLocation } from './apis/location/getLocation.js';
export { offLocationChange } from './apis/location/offLocationChange.js';
export { offLocationChangeError } from './apis/location/offLocationChangeError.js';
export { onLocationChange } from './apis/location/onLocationChange.js';
export { onLocationChangeError } from './apis/location/onLocationChangeError.js';
export { openLocation } from './apis/location/openLocation.js';
export { startLocationUpdate } from './apis/location/startLocationUpdate.js';
export { startLocationUpdateBackground } from './apis/location/startLocationUpdateBackground.js';
export { stopLocationUpdate } from './apis/location/stopLocationUpdate.js';
export { createAudioContext, createInnerAudioContext } from './apis/media/audio/index.js';
export { createCameraContext } from './apis/media/camera.js';
export { compressImage } from './apis/media/image/compressImage.js';
export { getImageInfo } from './apis/media/image/getImageInfo.js';
export { previewImage } from './apis/media/image/previewImage.js';
export { previewMedia } from './apis/media/image/previewMedia.js';
export { saveImageToPhotosAlbum } from './apis/media/image/saveImageToPhotosAlbum.js';
export { chooseImage } from './apis/media/image/chooseImage.js';
export { createLivePlayerContext } from './apis/media/live/index.js';
export { createMapContext } from './apis/media/map/index.js';
export { getRecorderManager } from './apis/media/recorder.js';
export { getVideoInfo } from './apis/media/video/getVideoInfo.js';
export { openVideoEditor } from './apis/media/video/openVideoEditor.js';
export { saveVideoToPhotosAlbum } from './apis/media/video/saveVideoToPhotosAlbum.js';
export { chooseVideo } from './apis/media/video/chooseVideo.js';
export { compressVideo } from './apis/media/video/compressVideo.js';
export { chooseMedia } from './apis/media/video/chooseMedia/index.js';
export { exitMiniProgram, navigateTo } from './apis/navigate/index.js';
export { downloadFile } from './apis/network/download.js';
export { addInterceptor, cleanInterceptors, request } from './apis/network/request/index.js';
export { uploadFile } from './apis/network/upload.js';
export { getAccountInfoSync } from './apis/open-api/account.js';
export { authorizeForMiniProgram } from './apis/open-api/authorize.js';
export { openChannelsUserProfile } from './apis/open-api/channels-live.js';
export { login } from './apis/open-api/login/login.js';
export { getSetting } from './apis/open-api/settings/getSetting.js';
export { openSetting } from './apis/open-api/settings/openSetting.js';
export { checkIsSoterEnrolledInDevice } from './apis/open-api/soter/checkIsSoterEnrolledInDevice.js';
export { checkIsSupportSoterAuthentication } from './apis/open-api/soter/checkIsSupportSoterAuthentication.js';
export { startSoterAuthentication } from './apis/open-api/soter/startSoterAuthentication.js';
export { requestSubscribeMessage } from './apis/open-api/subscribe-message.js';
export { getUserInfo } from './apis/open-api/user-info/getUserInfo.js';
export { requestPayment } from './apis/payment/requestPayment.js';
export { NativeDataCache, SyncCacheProxyHandler } from './apis/NativeApiSyncCacheProxy.js';
export { getShareInfo, hideShareMenu, showShareMenu } from './apis/share/index.js';
export { getBackgroundFetchData } from './apis/storage/getBackgroundFetchData.js';
export { setBackgroundFetchToken } from './apis/storage/setBackgroundFetchToken.js';
export { clearStorage, clearStorageSync, createBufferURL, createCacheManager, getStorage, getStorageInfo, getStorageInfoSync, getStorageSync, removeStorage, removeStorageSync, revokeBufferURL, setStorage, setStorageSync } from './apis/storage/index.js';
export { nextTick } from './apis/ui/custom-component.js';
export { disableAlertBeforeUnload, enableAlertBeforeUnload } from './apis/ui/interaction/index.js';
export { getMenuButtonBoundingClientRect } from './apis/ui/menu.js';
export { loadNavigationStyle, registerNavigationStyleHandler } from './apis/ui/navigation-bar/style-event.js';
export { hideHomeButton, hideNavigationBarLoading, setNavigationBarColor, setNavigationBarTitle, showNavigationBarLoading } from './apis/ui/navigation-bar/index.js';
export { setTopBarText } from './apis/ui/sticky.js';
export { createIntersectionObserver, createSelectorQuery } from './apis/wxml/index.js';
export { getAppInfo } from '@tarojs/taro-h5/dist/api/taro';

function safeStringify(obj) {
    try {
        return JSON.stringify(obj, null, 2);
    }
    catch (error) {
        // 在处理非法输入时捕获异常
        return String(obj); // 返回原始值的字符串表示形式
    }
}
function logFun(funObj, message) {
    return new Proxy(funObj, {
        apply(target, thisArg, args) {
            // eslint-disable-next-line no-console
            console.debug('%s, 参数为: %s', message, safeStringify(args));
            return target.apply(thisArg, args);
        },
    });
}
function logObj(obj) {
    return new Proxy(obj, {
        get(target, property) {
            if (typeof target[property] !== 'function') {
                const value = target[property];
                // @ts-ignore
                // eslint-disable-next-line no-console
                console.debug('调用 native %o property，结果为：%s ', property, safeStringify(value));
                return value;
            }
            return function (...args) {
                // @ts-ignore
                let log = `调用 native.${property} 方法`;
                if (args && args[0]) {
                    log = `${log}, 参数为${safeStringify(args[0])}`;
                    if (typeof args[0].success === 'function') {
                        // @ts-ignore
                        args[0].success = logFun(args[0].success, `${log} success回调结果`);
                    }
                    if (typeof args[0].fail === 'function') {
                        // @ts-ignore
                        args[0].fail = logFun(args[0].fail, `${log} fail回调结果`);
                    }
                }
                const result = target[property].apply(this, args);
                // eslint-disable-next-line no-console
                console.debug('%s, 返回值为: %s', log, safeStringify(result));
                return result;
            };
        },
    });
}
if (typeof window !== 'undefined' && process.env.NODE_ENV === 'development') {
    console.warn(`当前为development阶段，输出所有原生Api的调用日志。`);
    // @ts-ignore
    window.native = logObj(window.native || {});
}

export { taro as default };
