// @tarojs/plugin-platform-harmony-hybrid/dist/api/apis/device/bluetooth-ble/getBLEDeviceCharacteristics.js
import { shouldBeObject, getParameterError } from '../../utils/index.js';
import { MethodHandler } from '../../utils/handler.js';
import native from '../../NativeApi.js';

/**
 * 获取蓝牙设备某个服务中所有特征值
 *
 * @canUse getBLEDeviceCharacteristics
 * @__object [deviceId, serviceId]
 * @__success [characteristics]
 */
const getBLEDeviceCharacteristics = (options) => {
    const name = 'getBLEDeviceCharacteristics';
    return new Promise((resolve, reject) => {
        // options must be an Object
        const isObject = shouldBeObject(options);
        if (!isObject.flag) {
            const res = { errMsg: `${name}:fail ${isObject.msg}` };
            console.error(res.errMsg);
            return reject(res);
        }
        const { deviceId, serviceId, success, fail, complete } = options;
        const handle = new MethodHandler({ name, success, fail, complete });
        // options.deviceId must be string
        if (typeof deviceId !== 'string') {
            return handle.fail({
                errMsg: getParameterError({
                    para: 'deviceId',
                    correct: 'string',
                    wrong: deviceId,
                }),
            }, { resolve, reject });
        }
        if (typeof serviceId !== 'string') {
            return handle.fail({
                errMsg: getParameterError({
                    para: 'serviceId',
                    correct: 'string',
                    wrong: serviceId,
                }),
            }, { resolve, reject });
        }
        native.getBLEDeviceCharacteristics({
            deviceId: deviceId,
            serviceId: serviceId,
            success: (res) => {
                handle.success(res, { resolve, reject });
            },
            fail: (err) => {
                handle.fail(err, { resolve, reject });
            },
        });
    });
};

export { getBLEDeviceCharacteristics };
