// @tarojs/plugin-platform-harmony-hybrid/dist/api/apis/device/bluetooth-ble/readBLECharacteristicValue.js
import { shouldBeObject, getParameterError } from '../../utils/index.js';
import { MethodHandler } from '../../utils/handler.js';
import native from '../../NativeApi.js';

/**
 * 读取低功耗蓝牙设备的特征值的二进制数据值
 *
 * @canUse readBLECharacteristicValue
 * @__object [characteristicId, deviceId, serviceId]
 */
const readBLECharacteristicValue = (options) => {
    const name = 'readBLECharacteristicValue';
    return new Promise((resolve, reject) => {
        // options must be an Object
        const isObject = shouldBeObject(options);
        if (!isObject.flag) {
            const res = { errMsg: `${name}:fail ${isObject.msg}` };
            console.error(res.errMsg);
            return reject(res);
        }
        const { characteristicId, deviceId, serviceId, success, fail, complete } = options;
        const handle = new MethodHandler({ name, success, fail, complete });
        // options.characteristicId must be string
        if (typeof characteristicId !== 'string') {
            return handle.fail({
                errMsg: getParameterError({
                    para: 'characteristicId',
                    correct: 'string',
                    wrong: characteristicId,
                }),
            }, { resolve, reject });
        }
        if (typeof deviceId !== 'string') {
            return handle.fail({
                errMsg: getParameterError({
                    para: 'deviceId',
                    correct: 'string',
                    wrong: deviceId,
                }),
            }, { resolve, reject });
        }
        if (typeof serviceId !== 'string') {
            return handle.fail({
                errMsg: getParameterError({
                    para: 'serviceId',
                    correct: 'string',
                    wrong: serviceId,
                }),
            }, { resolve, reject });
        }
        native.readBLECharacteristicValue({
            characteristicId: characteristicId,
            deviceId: deviceId,
            serviceId: serviceId,
            success: (res) => {
                const result = {
                    /** 错误信息 */
                    errMsg: '',
                    /** 错误码 */
                    errCode: res[0],
                };
                handle.success(result, { resolve, reject });
            },
            fail: (err) => {
                handle.fail(err, { resolve, reject });
            },
        });
    });
};

export { readBLECharacteristicValue };
