// @tarojs/plugin-platform-harmony-hybrid/dist/api/apis/files/getFileInfo.js
import { shouldBeObject, getParameterError } from '../utils/index.js';
import { MethodHandler } from '../utils/handler.js';
import native from '../NativeApi.js';

/**
 * 关闭监听实时位置变化，前后台都停止消息接收
 *
 * @canUse getFileInfo
 * @__object [filePath, digestAlgorithm]
 * @__success [size, digest]
 */
const getFileInfo = (options) => {
    const name = 'getFileInfo';
    // options must be an Object
    const isObject = shouldBeObject(options);
    if (!isObject.flag) {
        const res = { errMsg: `${name}:fail ${isObject.msg}` };
        console.error(res.errMsg);
        return Promise.reject(res);
    }
    const { filePath, digestAlgorithm, success, fail, complete } = options;
    const handle = new MethodHandler({ name, success, fail, complete });
    // options.url must be String
    if (typeof filePath !== 'string') {
        return handle.fail({
            errMsg: getParameterError({
                para: 'filePath',
                correct: 'string',
                wrong: filePath,
            }),
        });
    }
    return new Promise((resolve, reject) => {
        native.getFileInfo({
            filePath: filePath,
            digestAlgorithm: digestAlgorithm,
            success: (res) => {
                const result = {
                    size: res.size,
                    digest: res.digest,
                    errMsg: res.errMsg,
                };
                handle.success(result, { resolve, reject });
            },
            fail: (err) => {
                const errRet = {
                    errMsg: err.errMsg,
                };
                handle.fail(errRet, { resolve, reject });
            },
        });
    });
};

export { getFileInfo };
