// @tarojs/plugin-platform-harmony-hybrid/dist/api/apis/files/removeSavedFile.js
import { shouldBeObject, getParameterError } from '../utils/index.js';
import { MethodHandler } from '../utils/handler.js';
import native from '../NativeApi.js';

/**
 * 新开页面打开文档，支持格式
 *
 * @canUse removeSavedFile
 * @__object [filePath]
 */
const removeSavedFile = (options) => {
    const name = 'removeSavedFile';
    // options must be an Object
    const isObject = shouldBeObject(options);
    if (!isObject.flag) {
        const res = { errMsg: `${name}:fail ${isObject.msg}` };
        console.error(res.errMsg);
        return Promise.reject(res);
    }
    const { filePath, success, fail, complete } = options;
    const handle = new MethodHandler({ name, success, fail, complete });
    // options.url must be String
    if (typeof filePath !== 'string') {
        return handle.fail({
            errMsg: getParameterError({
                para: 'filePath',
                correct: 'string',
                wrong: filePath,
            }),
        });
    }
    return new Promise((resolve, reject) => {
        native.removeSavedFile({
            filePath: filePath,
            success: (res) => {
                handle.success(res, { resolve, reject });
            },
            fail: (err) => {
                handle.fail(err, { resolve, reject });
            },
        });
    });
};

export { removeSavedFile };
