// @tarojs/plugin-platform-harmony-hybrid/dist/api/apis/base/getAppAuthorizeSetting/native.js
import native from '../../NativeApi.js';

/**
 * 获取APP授权设置
 *
 * @canUse getAppAuthorizeSetting
 * @__return
 * [albumAuthorized, bluetoothAuthorized, cameraAuthorized, locationAuthorized, locationReducedAccuracy,\
 * microphoneAuthorized, notificationAuthorized, phoneCalendarAuthorized]
 */
const getAppAuthorizeSetting = () => {
    const info = native.getAppAuthorizeSetting();
    // @ts-ignore
    const appAuthorizeSetting = {
        /** 允许微信使用相册的开关（仅 iOS 有效） */
        albumAuthorized: info.albumAuthorized,
        /** 允许微信使用蓝牙的开关（仅 iOS 有效） */
        bluetoothAuthorized: info.bluetoothAuthorized,
        /** 允许微信使用摄像头的开关 */
        cameraAuthorized: info.cameraAuthorized,
        /** 允许微信使用定位的开关 */
        locationAuthorized: info.locationAuthorized,
        /** 定位准确度。true 表示模糊定位，false 表示精确定位（仅 iOS 有效） */
        locationReducedAccuracy: info.locationAccuracy === 'reduced',
        /** 允许微信使用麦克风的开关 */
        microphoneAuthorized: info.microphoneAuthorized,
        /** 允许微信通知的开关 */
        notificationAuthorized: info.notificationAuthorized,
        /** 允许微信读写日历的开关 */
        phoneCalendarAuthorized: info.phoneCalendarAuthorized,
    };
    return appAuthorizeSetting;
};

export { getAppAuthorizeSetting };
